types/wordpress__data/index.d.ts

```diff
@@ -18,7 +18,8 @@ export { Action, combineReducers };
 //
 export type ResolveSelectorMap = Record<string, <T = unknown>(...args: readonly any[]) => Promise<T>>;
 export type SelectorMap = Record<string, <T = unknown>(...args: readonly any[]) => T>;
-export type DispatcherMap = Record<string, <T = void>(...args: readonly any[]) => T>;
+type EnsurePromise<T> = T extends Promise<any> ? T : Promise<T>;
+export type DispatcherMap = Record<string, <T = void>(...args: readonly any[]) => EnsurePromise<T>>;
```

Comment:
If we need to cater for the edge case where a dispatcher returns a synchronous result, we could add a second optional parameter to disable the promise wrapping... it starts to feel really heavy and complicated to use though:

```suggestion
type EnsurePromise<T> = T extends Promise<any> ? T : Promise<T>;
export type DispatcherMap = Record<string, <T = void, DoEnsurePromise extends boolean = true>(...args: readonly any[]) => DoEnsurePromise extends true ? EnsurePromise<T> : T>;
```

Reply:
> If we need to cater for the edge case where a dispatcher returns a synchronous result,

Thanks. Yeah, I'm trying to figure out if that's a possible situation from the source. An action creator can (I think) only return an action object or a thunk, and dispatching it (I think) returns either `Promise<void>` or whatever the thunk's return value is. It's not clear to me if a thunk's return value can be synchronous.

Reply:
At least with regular Redux Thunks I think they can return any arbitrary values, not just Promises (or at least I can't find anything in the code for Redux-Thunk that would suggest Promises are mandatory):

https://github.com/reduxjs/redux-thunk/blob/master/extend-redux.d.ts#L37-L41

Reply:
Good find. I can't find any examples of a non-Promise being returned in the [README](https://github.com/reduxjs/redux-thunk#readme) but [in the usage](https://redux.js.org/usage/writing-logic-thunks#returning-values-from-thunks) there is an example of returning a number. It reads,

> The most common use case for this is returning a promise from a thunk. 

but it does seem to support that use-case, which is kind of unfortunate for the changes here.

Even if we make the type support a Promise, without being able to guarantee it, we'd probably need a type guard on the dispatch's return value in order to safely await it, unless the type system can somehow deduce the return type of each action creator?

Reply:
`await 42` is valid JS, it just resolves to `42`. Fortunately for this use case await ["[r]eturns the fulfilled value of the promise, or the value itself if it's not a Promise.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#syntax). No need to conditionally check if something is actually a Promise before `await`ing it :slightly_smiling_face:

Reply:
TIL! But what if the consumer wanted to `then` the return value?

```js
// ... defining the action creators ...
const doSomething = () => () => 42;
// ... then later ...
const actions = useDispatch('something');
actions.doSomething().then( performFollowUp );
```

Reply:
> On the consuming side of the dispatcher it's up to the user to know whether it should be awaited or not. The internal implementation doesn't need to check is all I was saying.

I agree. I'm just wondering, if we implement your suggestion here (or something similar) to account for non-Promises to be returned, does that mean that any TS client code which tries to use `.then()` on the return value will have to include a type guard to satisfy the compiler?

Reply:
Only when the return value is typed as a non-promise (in which case I would think they wouldn't be `.then()`ing the value).

Maybe I am missing something from what you're asking. If someone is already `then`ing a non-Promise then it's a valid error that TS should catch. If they are trying to `then` something that isn't typed as a Promise but is actually a Promise then the signature for the method should be fixed to reflect that (either by fixing the inferred type or by explicitly typing it).

Reply:
I'm probably just confused. I guess here's what I'm trying to find out:

1. Should we use the code as currently exists in this PR? Or,
2. Should we use the suggestion you made here?

Reply:
I think while my suggestion is kind of cumbersome it does allow the thunks to be technically correct and allow for synchronous thunks (however rare they might be). It seems necessary, despite being ugly (maybe someone else can come up with a better approach for it :slightly_smiling_face:)